src/Metadata/Property/Factory/ConstructorPropertyMetadataFactory.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Metadata\Property\Factory;
+
+use ApiPlatform\Core\Api\OperationMethodResolverInterface;
+use ApiPlatform\Core\Exception\PropertyNotFoundException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+
+/**
+ * Set properties available in the constructor as writable.
+ *
+ * @author Maxime Veber <maxime.veber@nekland.fr>
+ */
+final class ConstructorPropertyMetadataFactory implements PropertyMetadataFactoryInterface
+{
+    /**
+     * @var PropertyMetadataFactoryInterface
+     */
+    private $decorated;
+
+    /**
+     * @var OperationMethodResolverInterface
+     */
+    private $operationMethodResolver;
+
+    public function __construct(OperationMethodResolverInterface $operationMethodResolver, PropertyMetadataFactoryInterface $decorated = null)
```

Comment:
Why would the decorated object be `null` ?

Reply:
so that it can be used without decoration

Reply:
IMO it's either you decorate something or you don't, Not both at the same time (S from Solid, anyone ?).

Having a basic decorator that just returns a `new PropertyMetadata()` would be simpler.

Reply:
@Taluu loaders can be chained in any order (and sometimes you have to change the order, to change the priority). The first registered loader will not be a decorator, but you cannot guess which one will be the first when writing the code (it depends of the config).

Reply:
Then it means it is the wrong design pattern for these. Shouldn't a strategy be better for that ? of course, nothing prevents some elements in this pattern to be decorators...

Reply:
Or a chain of responsability would also do it

Reply:
It was tried, and they introduces a lof of unneeded complexity. They have been a lot of iteration on this part, I don't want to change things again. It's mostly internal, and it works well as is.

Reply:
+using a decorator makes it very easy to configure with the Symfony DI component since it supports service decoration and priority. Using a different pattern such as chain (as it was in API Platform 1) will require to add a lot of specific code to have the same flexibility.